common/src/retry.rs

```diff
@@ -16,62 +16,275 @@
 //! }
 //! ```
 
+use crate::time::{Duration, Instant};
+use arc_swap::ArcSwap;
 use rand::Rng;
+use std::sync::{
+    atomic::{AtomicBool, AtomicUsize, Ordering},
+    Arc,
+};
 
 pub struct NotSpecialized;
 
 /// Specifies which errors are retryable.
 /// All Errors are not retryable by-default.
 pub trait RetryableError<SP = NotSpecialized>: std::error::Error {
     fn is_retryable(&self) -> bool;
+
+    /// If the global scope this `Retry` operates in needs to be
+    /// backed off on an error (e.g. a Rate Limit) this should return `true`
+    fn needs_cooldown(&self) -> bool {
+        false
+    }
+}
+
+impl<T> RetryableError for &'_ T
+where
+    T: RetryableError,
+{
+    fn is_retryable(&self) -> bool {
+        (**self).is_retryable()
+    }
+
+    fn needs_cooldown(&self) -> bool {
+        (**self).needs_cooldown()
+    }
+}
+
+impl<E: RetryableError> RetryableError for Box<E> {
+    fn is_retryable(&self) -> bool {
+        (**self).is_retryable()
+    }
+
+    fn needs_cooldown(&self) -> bool {
+        (**self).needs_cooldown()
+    }
 }
 
 /// Options to specify how to retry a function
-#[derive(Debug, PartialEq, Eq, Copy, Clone)]
-pub struct Retry {
+#[derive(Debug, Clone)]
+pub struct Retry<S = ExponentialBackoff, C = ()> {
     retries: usize,
-    duration: core::time::Duration,
-    // The amount to multiply the duration on each subsequent attempt
-    multiplier: u32,
-    max_jitter_ms: usize,
+    strategy: S,
+    /// global cooldown for this retry strategy
+    cooldown: C,
+    /// whether we are currently in a cooldown period
+    is_cooling: Arc<AtomicBool>,
+    /// since when have we been cooling
+    cooling_since: Arc<ArcSwap<crate::time::Instant>>,
```

Comment:
Seeing this `ArcSwap` lib for the first time. That's useful.

Reply:
Nit: would probably be better to use SystemTime here and just store the NS since epoch in an atomic.